RavenDB-19951 Introduce TOTP support for browser generated requests

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19951 

### Additional description

This allows an administrator to define a TOTP key for a certificate. A certificate with such a key will _not_ work unless the user will utilize a dedicated endpoint to provide a TOTP code. At which point, the certificate will be valid for a set duration. Afterward, a new TOTP key would be required.

This is meant for administrators certificates, not for applications to use.

The current version is a draft, to show the overall solution. Some notes about how this works:

* The TOTP provided is stored globally, but needs to be provided separately for each node in the cluster. We can make is cluster wide configuration, but I'm not sure if that matters or is even desirable.
* The limit happens when the actual connection is established. If we have a 1 minute validity period, but the connection is still being used, it'll keep on working.

UX concerns:

* We need some Studio work to generate the key, render it, etc.
* I didn't provide a way to extract the key out of the system. Not sure how important that is. 
* Need to provide a way to *update* this? What about updating just the validity period? 

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works


### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio

ayende committed:
RavenDB-19951 Introduce TOTP support for browser generated requests

ml054 commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -99,6 +100,8 @@ static RavenServer()
         private readonly ExternalCertificateValidator _externalCertificateValidator;
         internal readonly JsonContextPool _tcpContextPool;
 
+        private readonly ConcurrentDictionary<string, DateTime> _twoFactoAuthTimeByCertThumbprintExpiry = new();
```
typo: twoFactoAuth

ml054 commented:
I've spoke with @ppekrol and this is what we end up with:

Can we add additional security layer and bind either:
a) include client ip address in authorization dictionary 
b) send cookie with some permit (bearer token)

But choosing this option we protect again following atack:
1) Mallory gets Alice's private key
2) Mallory can't access resource since isn;t available in permit dict
3) Alice access studio and provides 2fa
4) Now Mallory can access studio within 2h.

ayende commented:
We can add a limit per IP, but I wonder if this would be a good idea. 

That said, the whole *point* here is to allow multiple connections. IP limit handles this for different tabs, etc.

But what about, sign up via my phone, then access via my computer?

About bearer token. To handle CSRF, we can have the studio get a value and always send that in headers for all auth request, no?
That would prevent the CSRF scenario?

ayende committed:
RavenDB-19951 Fixing test

ayende commented:
I added support for limiting per IP, but did it in such a way that the user can chose if to do that.

Thinking more about this, we *have* to allow the 2FA to be independent. That is, to allow it to run without a cluster (since we can't assume that we'll have a healthy cluster when we try).

That means that the question on sending to the cluster is now a usability issue. Thoughts?

ppekrol commented:
> I added support for limiting per IP, but did it in such a way that the user can chose if to do that.
> 
> Thinking more about this, we _have_ to allow the 2FA to be independent. That is, to allow it to run without a cluster (since we can't assume that we'll have a healthy cluster when we try).
> 
> That means that the question on sending to the cluster is now a usability issue. Thoughts?

1. IMO using Bearer Token should be enough and best solution, we do not need any additional IP restrictions.
2. Why can't we assume that in order to setup 2FA we need to have a working cluster?

ayende commented:
1/ How would you enable the bearer token? How would you be able to run this if you are using shell scripts, etc? I wouldn't want to force them to do that. We had *bad* experience with `OAuth` in the past in that scenario.
2/ Consider the case where we have a bad cluster. An admin wants to log in and figure out what is going on. They cannot do that with 2FA, however, since that is a cluster operation.

ppekrol commented:
1. I think we do have misunderstanding here what is Bearer Token. And shell script is an app, so I would assume that this would require 'app' certificate, so without 2FA.
2. Setting up 2FA should be a cluster operation IMO, but not validating it?
3. I think it would be best to have a call with me and @ml054

ayende commented:
An admin wants to run a backup manually. They have a script that does this already.
Or they want to run a tool to create indexes, and they want to use *their* certificate for that.
It would fail with 2FA, but it can work if we auth, then we have the door open, for a certain period of time. No need to add additional behavior.

Bearer token for me is a separate issue for Studio requests, not normal ones.

Setting a 2FA is a cluster with operation, yes, like anything with certs.
Validating it doesn't *require* a cluster connection. I think we can send the auth to the rest of the cluster, for sure, but it must also be handled and managed when there is no cluster availabl.e

ayende committed:
RavenDB-19951
- Added support for cookie & header token for CSRF alongside 2FA configurations
- Allow to provide multiple IPs for 2FA scenarios

ml054 commented on src/Raven.Server/Routing/RequestRouter.cs:
```diff
@@ -140,25 +144,74 @@ internal async ValueTask<(bool Authorized, RavenServer.AuthenticationStatus Stat
+        {
+            if (context.Request.Cookies.TryGetValue(TwoFactorAuthentication.CookieName, out var cookieStr) == false)
+            {
+                msg = "Missing the csrf-cookie in the request";
```
do we want to use TwoFactorAuthentication.CookieName here?

ayende replied:
Fixed

ayende committed:
RavenDB-19951 Using constant in message string

ppekrol commented:
Done here: https://github.com/ravendb/ravendb/pull/17851
